core/src/surplus.ts

```diff
@@ -0,0 +1,88 @@
+import type { SurplusAuction, SurplusAuctionUnexistant, SurplusAuctionBaseData } from './types';
+import { getEarliestDate } from './helpers/getEarliestDate';
+import BigNumber from './bignumber';
+import getContract from './contracts';
+import { Contract } from 'ethers';
+import getNetworkDate from './date';
+import { RAD, WAD } from './constants/UNITS';
+
+const getSurplusAuctionLastIndex = async (contract: Contract): Promise<number> => {
+    const auctionsQuantityBinary = await contract.kicks();
+    return new BigNumber(auctionsQuantityBinary._hex).toNumber();
+};
+
+const getAuctionState = async (network: string, earliestEndDate: Date, greatestBid: number) => {
+    const isBidExpired = (await getNetworkDate(network)) > earliestEndDate;
+    const haveBids = greatestBid === 0;
+    if (haveBids) {
+        if (isBidExpired) {
+            return 'ready-for-collection';
+        }
+        return 'have-bids';
+    }
+    if (isBidExpired) {
+        return 'requires-restart';
+    }
+    return 'just-started';
+};
+
+export const fetchSurplusAuctionByIndex = async function (
+    network: string,
+    auctionIndex: number
+): Promise<SurplusAuction | SurplusAuctionUnexistant> {
+    const contract = await getContract(network, 'MCD_FLAP');
+    const auctionData = await contract.bids(auctionIndex);
+    const isAuctionDeleted = new BigNumber(auctionData.end).eq(0);
+    const baseAuctionInfo: SurplusAuctionBaseData = {
+        network,
+        id: auctionIndex,
+    };
+
+    if (isAuctionDeleted) {
+        const auctionLastIndex = await getSurplusAuctionLastIndex(contract);
+        if (auctionLastIndex < auctionIndex) {
+            throw new Error('No active auction exists this id');
+        }
+        return { ...baseAuctionInfo, state: 'collected' };
+    }
+
+    const auctionEndDate = new Date(auctionData.end * 1000);
+    const bidEndDate = new Date(auctionData.tic * 1000);
+    const earliestEndDate = getEarliestDate(auctionEndDate, bidEndDate);
+    const state = await getAuctionState(network, earliestEndDate, auctionData.tic);
+
+    return {
+        ...baseAuctionInfo,
+        earliestEndDate,
+        bidAmountMKR: new BigNumber(auctionData.bid._hex).div(WAD),
+        receiveAmountDAI: new BigNumber(auctionData.lot._hex).div(RAD),
+        receiverAddress: auctionData.guy,
+        auctionEndDate,
+        bidEndDate,
+        state,
+    };
+};
+
+const getActiveSurplusAuctionOrUndefined = async (network: string, auctionIndex: number) => {
+    const auction = await fetchSurplusAuctionByIndex(network, auctionIndex);
```

Comment:
Can there be a case that it throws an error?

Reply:
Within this context there're no errors i am aware of that we expect and have to handle.

Reply:
1. `fetchSurplusAuctionByIndex` throws an error
2. underlying provider may also throw and error on the network timeout

Reply:
> Usually (to my xp) there's something one has to do if the error rises.

Yes, my question was, if there is an edge case where this error expected